---
title: "700-structure"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Let's import our genindMAF and filter our SNPs in LD

```{r import, warning=FALSE, message=FALSE}
library("tidyverse")
library("poppr")
library("magrittr")

load(file="outputs/400/genMaf.rda")
```

### Calculate LD
```{r LD, warning=FALSE, message=FALSE, eval=FALSE}
genc<-as.genclone(genMaf)
ia(gid = genc, sample = 100)
```
Test reveals.... No linkage?
|======================================================================|100% ~0 s remaining     
          Ia         p.Ia        rbarD         p.rD   
208.06405006   0.00990099   0.03349214   0.00990099   

We need a STRUCTURE file
```{r STR, warning=FALSE, message=FALSE, eval=FALSE}
library(adegenet)
library(pegas)
library(poppr)

df <- as.loci(genMac, allele.sep="\t") ###CONVERSION OF GENIND OBJECT TO LOCI OBJECT####
write.loci(df, file = "outputs/700/175.str", loci.sep ="\t", quote = FALSE,
           allele.sep ="\t", na ="-9\t-9", col.names = FALSE) ###WRITE STRUCTURE DATA FILE####
```

Getting ParallelStructure to run?

```{r anotherFormat, message=FALSE, warning=FALSE}
# Function to export to STRUCTURE format from genind object.
# genind objects are created in the R package adegenet.  The function below is an R function.
# Lindsay V. Clark, 26 July 2015

# obj: genind object
# file: file name to write
# pops: whether to include population info in the file
# Function is flexible with regards to ploidy, although genotypes are
# considered to be unambiguous.
# Missing data must be recorded as NA in obj@tab.

# example use: 
# data(nancycats)
# genind2structure(nancycats, file="nancy_structure.txt", pops=TRUE)

genind2structure <- function(obj, file="", pops=FALSE){
  if(!"genind" %in% class(obj)){
    warning("Function was designed for genind objects.")
  }
  
  # get the max ploidy of the dataset
  pl <- max(obj@ploidy)
  # get the number of individuals
  S <- adegenet::nInd(obj)
  # column of individual names to write; set up data.frame
  tab <- data.frame(ind=rep(indNames(obj), each=pl))
  # column of pop ids to write
  if(pops){
    popnums <- 1:adegenet::nPop(obj)
    names(popnums) <- as.character(unique(adegenet::pop(obj)))
    popcol <- rep(popnums[as.character(adegenet::pop(obj))], each=pl)
    tab <- cbind(tab, data.frame(pop=popcol))
  }
  loci <- adegenet::locNames(obj) 
  # add columns for genotypes
  tab <- cbind(tab, matrix(-9, nrow=dim(tab)[1], ncol=adegenet::nLoc(obj),
                           dimnames=list(NULL,loci)))
  
  # begin going through loci
  for(L in loci){
    thesegen <- obj@tab[,grep(paste("^", L, "\\.", sep=""), 
                              dimnames(obj@tab)[[2]]), 
                        drop = FALSE] # genotypes by locus
    al <- 1:dim(thesegen)[2] # numbered alleles
    for(s in 1:S){
      if(all(!is.na(thesegen[s,]))){
        tabrows <- (1:dim(tab)[1])[tab[[1]] == indNames(obj)[s]] # index of rows in output to write to
        tabrows <- tabrows[1:sum(thesegen[s,])] # subset if this is lower ploidy than max ploidy
        tab[tabrows,L] <- rep(al, times = thesegen[s,])
      }
    }
  }
  
  # export table
  write.table(tab, file=file, sep="\t", quote=FALSE, row.names=FALSE)
}

#genind2structure(kenai, file="outputs/700/genMaf-alt-str.txt", pops=TRUE)
```

Example file:  
016	1	1	1	4	4	4	4	2	2	2	2	4	4	4	4  
017	1	1	1	4	4	4	4	2	2	2	3	4	4	4	3  
018	1	1	1	4	4	4	4	2	2	2	2	4	4	4	4  
019	1	1	1	4	4	4	4	2	2	2	3	4	4	4	3  
033	1	1	1	4	4	4	4	2	2	2	2	4	4	4	4  
034	2	1	1	4	4	4	4	2	2	2	2	4	4	4	4  
035	2	1	1	4	4	4	4	2	2	2	2	4	4	4	4  
036	2	1	1	4	4	4	4	2	2	2	2	4	4	4	4  
037	2	1	1	4	4	4	4	2	2	2	2	4	4	4	4  
038	2	1	1	4	4	4	4	2	2	2	2	4	4	4	4  

Create a header, add pops, and do some fiddling to make numeric
```{sh header, eval=FALSE}
for i in {1..6336}; do echo Locus$i | perl -pe 's/\n/\t/' >> outputs/700/header.txt; done;
cat structure.str | cut -f 1,2 | awk '{print $2 $1}' >> indNames.txt
cat structure.str | perl -pe 's/A/1/g' | perl -pe 's/G/2/g' | perl -pe 's/C/3/g' | perl -pe 's/T/4/g' > out.txt

cat structure.str | perl -pe 's/(1|2|3|4)(1|2|3|4)/$1\t$2/g' > str2.txt
cat names.txt | perl -pe 's/^\w+(\d\d\d)/$1/g' > numeric.txt
#see str3.txt
paste namespops.txt str2.txt > str3.txt

```

##Read in Structure output
```{r, message=FALSE, warning=FALSE}
qdata<-read_delim("outputs/700/results_job_T4_q", col_names = FALSE, delim=" ")
qdata$Individual<-gsub("\\s+", "", qdata$X1)
qdata$Population<-gsub("\\s+", "", qdata$X2)
qdata<-qdata %>% rename(Q1=X3, Q2=X4, Q3=X5) %>% select(Population, Individual, Q1, Q2, Q3)

pops<-read_tsv("outputs/700/popList.txt", col_names = FALSE)
qdata$Pop<-pops$X1

df<-qdata %>% select(Population, Pop, Individual, Q1, Q2, Q3) %>% gather(key=Ancestry, value=Q, 4:6)

ggplot(df)+geom_col(aes(x=Individual,y=Q, fill=Ancestry))+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  scale_fill_brewer(palette ="Spectral")+
  facet_wrap(. ~ Pop, ncol=4, scales="free_x")

```

```{r, message=FALSE, warning=FALSE}
qdata<-read_delim("outputs/700/results_job_T7_q", col_names = FALSE, delim=" ")
qdata$Individual<-gsub("\\s+", "", qdata$X1)
qdata$Population<-gsub("\\s+", "", qdata$X2)
qdata<-qdata %>% rename(Q1=X3, Q2=X4, Q3=X5, Q4=X6) %>% select(Population, Individual, Q1, Q2, Q3, Q4)

pops<-read_tsv("outputs/700/popList.txt", col_names = FALSE)
qdata$Pop<-pops$X1

df<-qdata %>% select(Population, Pop, Individual, Q1, Q2, Q3, Q4) %>% gather(key=Ancestry, value=Q, 4:7)

ggplot(df)+geom_col(aes(x=Individual,y=Q, fill=Ancestry))+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  scale_fill_brewer(palette ="Spectral")+
  facet_wrap(. ~ Pop, ncol=4, scales="free_x")

```

```{r, message=FALSE, warning=FALSE}
qdata<-read_delim("outputs/700/ParallelStructure2/results_job_T10_q", col_names = FALSE, delim=" ")
qdata$Individual<-gsub("\\s+", "", qdata$X1)
qdata$Population<-gsub("\\s+", "", qdata$X2)
qdata<-qdata %>% rename(Q1=X3, Q2=X4, Q3=X5, Q4=X6, Q5=X7) %>% select(Population, Individual, Q1, Q2, Q3, Q4, Q5)

pops<-read_tsv("outputs/700/popList.txt", col_names = FALSE)
qdata$Pop<-pops$X1

df<-qdata %>% select(Population, Pop, Individual, Q1, Q2, Q3, Q4, Q5) %>% gather(key=Ancestry, value=Q, 4:8)

ggplot(df)+geom_col(aes(x=Individual,y=Q, fill=Ancestry))+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  scale_fill_brewer(palette ="Spectral")+
  facet_wrap(. ~ Pop, ncol=4, scales="free_x")

```