---
title: "900-sfs-estimation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## SFS

For estimating thetas and neutrality stats, we are going to use a folded SFS.  

I'll get some test data.  

`Macs-MBP-2:data mac$ mkdir test`  
`Macs-MBP-2:data mac$ cd test/`  
`Macs-MBP-2:test mac$ wget http://popgen.dk/software/download/angsd/bams.tar.gz`  
`Macs-MBP-2:test mac$ tar -xvf bams.tar.gz`  

Index and bamlist.  
`Macs-MBP-2:test mac$ for i in bams/*.bam;do samtools index $i;done`  
`Macs-MBP-2:test mac$ ls bams/*.bam > bam.filelist`  

Get reference
`Macs-MBP-2:pike-time mac$ cd genome/`  
`Macs-MBP-2:genome mac$ ls`  
`Macs-MBP-2:genome mac$ wget http://popgen.dk/software/download/angsd/hg19ancNoChr.fa.gz`  
`Macs-MBP-2:genome mac$ mv hg19ancNoChr.fa.gz chimpHg19.fa.gz`  
`Macs-MBP-2:genome mac$ samtools faidx chimpHg19.fa.gz`  

Prepare for output.  
`Macs-MBP-2:pike-time mac$ mkdir outputs/900/`  

Now I can follow: http://www.popgen.dk/angsd/index.php/Thetas,Tajima,Neutrality_tests   
Except, I will deviate and calculate a folded SFS.  

  __1__ Estimate an site frequency spectrum. Output is out.sfs file. This is what is being used as the -pest argument in step2.  
  __2__ Calculate per-site thetas. Output is a .thetas.idx/.thetas.gz files. This contains the binary per-site estimates of the thetas.  
  __3__ Calculate neutrality tests statistics. Output is a .thetas.idx.pestPG file.   
  
Here are some steps I used before:   

`angsd -bam $bamlist -out out -doSaf 1 -doPost 2 -GL 2 -doMaf 2 -doMajorMinor 1 - minMaf 0.05 -minInd $minind -minMapQ 30 -minQ 30 -SNP pval 2e-3 -anc $reference -P 5 -fold 1`  
`realSFS out.saf.idx -P 5 > out.sfs`   
`angsd -bam $bamlist -out out -doThetas 1 -doSaf 1 -pest out.sfs -anc $reference -GL 2 -doMaf 2 -doMajorMinor 1 -minMaf 0.05 -minInd $minind -minMapQ 30 -minQ 30 -SNP pval 2e-3 -P 5 -fold 1`  
And that worked...  
#### Step 1 Folded SFS  

Copying from chinook-wgs/102-generate-folded-SFS.sh
```{sh, eval=FALSE}
/Users/mac/angsd/angsd -bam data/test/bam.filelist -doSaf 1 -doPost 2 -GL 2 -doMaf 2 -doMajorMinor 1 -minMaf 0.05 -minMapQ 30 -minQ 30 -SNP_pval 2e-3 -anc ./genome/chimpHg19.fa.gz  -fold 1 -nThreads 2 -out outputs/900/test 

/Users/mac/angsd/misc/realSFS outputs/900/test.saf.idx -maxIter 100 -P 2 -fold 1 > outputs/900/test-realSFS.stdout 2> outputs/900/test-realSFS.stderr
```

#### Step 2 Per-Site Thetas  

lols, this works.
```{sh, eval=FALSE}
/Users/mac/angsd/angsd -bam data/test/bam.filelist -out outputs/900/per-site -doThetas 1 -doSaf 1 -pest outputs/900/test-realSFS.stdout -fold 1 -GL 2 -anc genome/chimpHg19.fa.gz
```

What do they look like?

```{sh, eval=FALSE}
/Users/mac/angsd/misc/thetaStat print outputs/900/per-site.thetas.idx > outputs/900/per-site.thetas.idx.txt
```

```{r, warning=FALSE, message=FALSE}
library(tidyverse)
data<-read_tsv("outputs/900/per-site.thetas.idx.txt")

```
```{r, warning=FALSE, message=FALSE}
ggplot(data)+geom_boxplot(aes(x=`#Chromo`, y=Watterson))

mean(data$Watterson)
median(data$Watterson)

```

